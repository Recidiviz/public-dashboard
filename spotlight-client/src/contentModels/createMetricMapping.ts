// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { assertNever } from "assert-never";
import { MetricTypeIdList, TenantContent, TenantId } from "../contentApi/types";
import {
  parolePopulationCurrent,
  parolePopulationHistorical,
  paroleProgramParticipationCurrent,
  paroleRevocationReasons,
  paroleSuccessRateDemographics,
  paroleSuccessRateMonthly,
  prisonAdmissionReasons,
  prisonPopulationCurrent,
  prisonPopulationHistorical,
  prisonReleaseTypes,
  prisonStayLengths,
  probationPopulationCurrent,
  probationPopulationHistorical,
  probationProgramParticipationCurrent,
  probationRevocationReasons,
  probationSuccessRateDemographics,
  probationSuccessRateMonthly,
  recidivismRateAllFollowup,
  recidivismRateConventionalFollowup,
  sentencePopulationCurrent,
  sentenceTypesCurrent,
} from "../metricsApi";
import { MetricMapping } from "./types";
import DemographicsByCategoryMetric from "./DemographicsByCategoryMetric";
import HistoricalPopulationBreakdownMetric from "./HistoricalPopulationBreakdownMetric";
import PopulationBreakdownByLocationMetric from "./PopulationBreakdownByLocationMetric";
import ProgramParticipationCurrentMetric from "./ProgramParticipationCurrentMetric";
import RecidivismRateMetric from "./RecidivismRateMetric";
import SentenceTypeByLocationMetric from "./SentenceTypeByLocationMetric";
import SupervisionSuccessRateDemographicsMetric from "./SupervisionSuccessRateDemographicsMetric";
import SupervisionSuccessRateMonthlyMetric from "./SupervisionSuccessRateMonthlyMetric";
import { ERROR_MESSAGES } from "../constants";
import { NOFILTER_KEY, TOTAL_KEY } from "../demographics";
import { colors } from "../UiLibrary";

type MetricMappingFactoryOptions = {
  localityLabelMapping: TenantContent["localities"];
  metadataMapping: TenantContent["metrics"];
  tenantId: TenantId;
};
/**
 * Factory function for converting a mapping of content objects by metric ID
 * to a mapping of Metric instances by metric ID. Creating the entire mapping at once
 * ensures that each ID maps to the proper Metric type without requiring further
 * type guarding on the part of consumers.
 */
export default function createMetricMapping({
  localityLabelMapping,
  metadataMapping,
  tenantId,
}: MetricMappingFactoryOptions): MetricMapping {
  const metricMapping: MetricMapping = new Map();

  // to maintain type safety we iterate through all of the known metrics;
  // iterating through the metadata object's keys widens the type to `string`,
  // which prevents us from guaranteeing exhaustiveness at the type level
  MetricTypeIdList.forEach((metricType) => {
    // not all metrics are required; metadata object is the source of truth
    // for which metrics to include
    const metadata = metadataMapping[metricType];
    if (!metadata) {
      return;
    }

    // messages for errors we may hit at various points below
    const localityContentError = `${ERROR_MESSAGES.missingRequiredContent} (locality labels)`;
    const totalLabelError = `${ERROR_MESSAGES.missingRequiredContent} (total population label)`;

    // this big ol' switch statement ensures that the type ID string union is properly narrowed,
    // allowing for 1:1 correspondence between the ID and the typed Metric instance.
    switch (metricType) {
      case "SentencePopulationCurrent":
        if (!localityLabelMapping?.Sentencing)
          throw new Error(localityContentError);

        if ("totalLabel" in metadata)
          metricMapping.set(
            metricType,
            new PopulationBreakdownByLocationMetric({
              ...metadata,
              id: metricType,
              tenantId,
              defaultDemographicView: NOFILTER_KEY,
              defaultLocalityId: TOTAL_KEY,
              localityLabels: localityLabelMapping.Sentencing,
              dataTransformer: sentencePopulationCurrent,
              sourceFileName: "sentence_type_by_district_by_demographics",
            })
          );
        else throw new Error(totalLabelError);

        break;
      case "SentenceTypesCurrent":
        if (!localityLabelMapping?.Sentencing)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SentenceTypeByLocationMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Sentencing,
            dataTransformer: sentenceTypesCurrent,
            sourceFileName: "sentence_type_by_district_by_demographics",
          })
        );
        break;
      case "PrisonPopulationCurrent":
        if (!localityLabelMapping?.Prison)
          throw new Error(localityContentError);

        if ("totalLabel" in metadata)
          metricMapping.set(
            metricType,
            new PopulationBreakdownByLocationMetric({
              ...metadata,
              id: metricType,
              tenantId,
              defaultDemographicView: NOFILTER_KEY,
              defaultLocalityId: TOTAL_KEY,
              localityLabels: localityLabelMapping.Prison,
              dataTransformer: prisonPopulationCurrent,
              sourceFileName:
                "incarceration_population_by_facility_by_demographics",
            })
          );
        else throw new Error(totalLabelError);
        break;
      case "ProbationPopulationCurrent":
        if (!localityLabelMapping?.Probation)
          throw new Error(localityContentError);
        if ("totalLabel" in metadata)
          metricMapping.set(
            metricType,
            new PopulationBreakdownByLocationMetric({
              ...metadata,
              id: metricType,
              tenantId,
              defaultDemographicView: NOFILTER_KEY,
              defaultLocalityId: TOTAL_KEY,
              localityLabels: localityLabelMapping.Probation,
              dataTransformer: probationPopulationCurrent,
              sourceFileName:
                "supervision_population_by_district_by_demographics",
            })
          );
        else throw new Error(totalLabelError);
        break;
      case "ParolePopulationCurrent":
        if (!localityLabelMapping?.Parole)
          throw new Error(localityContentError);
        if ("totalLabel" in metadata)
          metricMapping.set(
            metricType,
            new PopulationBreakdownByLocationMetric({
              ...metadata,
              id: metricType,
              tenantId,
              defaultDemographicView: NOFILTER_KEY,
              defaultLocalityId: TOTAL_KEY,
              localityLabels: localityLabelMapping.Parole,
              dataTransformer: parolePopulationCurrent,
              sourceFileName:
                "supervision_population_by_district_by_demographics",
            })
          );
        else throw new Error(totalLabelError);
        break;
      case "PrisonPopulationHistorical":
        metricMapping.set(
          metricType,
          new HistoricalPopulationBreakdownMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: prisonPopulationHistorical,
            sourceFileName: "incarceration_population_by_month_by_demographics",
          })
        );
        break;
      case "ProbationPopulationHistorical":
        metricMapping.set(
          metricType,
          new HistoricalPopulationBreakdownMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: probationPopulationHistorical,
            sourceFileName: "supervision_population_by_month_by_demographics",
          })
        );
        break;
      case "ParolePopulationHistorical":
        metricMapping.set(
          metricType,
          new HistoricalPopulationBreakdownMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: parolePopulationHistorical,
            sourceFileName: "supervision_population_by_month_by_demographics",
          })
        );
        break;
      case "ProbationProgrammingCurrent":
        if (!localityLabelMapping?.Probation)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new ProgramParticipationCurrentMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: undefined,
            defaultLocalityId: NOFILTER_KEY,
            localityLabels: localityLabelMapping.Probation,
            dataTransformer: probationProgramParticipationCurrent,
            sourceFileName: "active_program_participation_by_region",
          })
        );
        break;
      case "ParoleProgrammingCurrent":
        if (!localityLabelMapping?.Parole)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new ProgramParticipationCurrentMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: undefined,
            defaultLocalityId: NOFILTER_KEY,
            localityLabels: localityLabelMapping.Parole,
            dataTransformer: paroleProgramParticipationCurrent,
            sourceFileName: "active_program_participation_by_region",
          })
        );
        break;
      case "ProbationSuccessHistorical":
        if (!localityLabelMapping?.Probation)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SupervisionSuccessRateMonthlyMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: undefined,
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Probation,
            dataTransformer: probationSuccessRateMonthly,
            sourceFileName: "supervision_success_by_month",
          })
        );
        break;
      case "ParoleSuccessHistorical":
        if (!localityLabelMapping?.Parole)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SupervisionSuccessRateMonthlyMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: undefined,
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Parole,
            dataTransformer: paroleSuccessRateMonthly,
            sourceFileName: "supervision_success_by_month",
          })
        );
        break;
      case "ProbationSuccessAggregate":
        if (!localityLabelMapping?.Probation)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SupervisionSuccessRateDemographicsMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Probation,
            dataTransformer: probationSuccessRateDemographics,
            sourceFileName: "supervision_success_by_period_by_demographics",
          })
        );
        break;
      case "ParoleSuccessAggregate":
        if (!localityLabelMapping?.Parole)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SupervisionSuccessRateDemographicsMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Parole,
            dataTransformer: paroleSuccessRateDemographics,
            sourceFileName: "supervision_success_by_period_by_demographics",
          })
        );
        break;
      case "ProbationRevocationsAggregate":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: probationRevocationReasons,
            sourceFileName:
              "supervision_revocations_by_period_by_type_by_demographics",
          })
        );
        break;
      case "ParoleRevocationsAggregate":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: paroleRevocationReasons,
            sourceFileName:
              "supervision_revocations_by_period_by_type_by_demographics",
          })
        );
        break;
      case "PrisonAdmissionReasonsCurrent":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: prisonAdmissionReasons,
            sourceFileName: "incarceration_population_by_admission_reason",
          })
        );
        break;
      case "PrisonReleaseTypeAggregate":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: prisonReleaseTypes,
            sourceFileName: "incarceration_releases_by_type_by_period",
          })
        );
        break;
      case "PrisonRecidivismRateHistorical":
        metricMapping.set(
          metricType,
          new RecidivismRateMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: recidivismRateAllFollowup,
            sourceFileName: "recidivism_rates_by_cohort_by_year",
          })
        );
        break;
      case "PrisonRecidivismRateSingleFollowupHistorical":
        metricMapping.set(
          metricType,
          new RecidivismRateMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            followUpYears: 3,
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: recidivismRateConventionalFollowup,
            sourceFileName: "recidivism_rates_by_cohort_by_year",
          })
        );
        break;
      case "PrisonStayLengthAggregate":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: prisonStayLengths,
            sourceFileName: "incarceration_lengths_by_demographics",
            color: colors.dataVizNamed.get("teal"),
          })
        );
        break;
      default:
        assertNever(metricType);
    }
  });

  return metricMapping;
}
