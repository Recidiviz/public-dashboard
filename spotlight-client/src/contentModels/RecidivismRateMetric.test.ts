// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { runInAction, when } from "mobx";
import { DemographicView } from "../demographics";
import { reactImmediately } from "../testUtils";
import createMetricMapping from "./createMetricMapping";
import RecidivismRateMetric from "./RecidivismRateMetric";
import contentFixture from "./__fixtures__/tenant_content_exhaustive";

const testTenantId = "US_ND";
const testMetricId = "PrisonRecidivismRateSingleFollowupHistorical";
const testMetadataMapping = {
  [testMetricId]: contentFixture.metrics[testMetricId],
};

function getTestMetric() {
  return createMetricMapping({
    localityLabelMapping: contentFixture.localities,
    metadataMapping: testMetadataMapping,
    tenantId: testTenantId,
  }).get(testMetricId) as RecidivismRateMetric;
}

async function getPopulatedMetric() {
  const metric = getTestMetric();

  metric.populateAllRecords();

  await when(() => Boolean(metric.records));

  return metric;
}

describe("single followup period", () => {
  test("total data", async () => {
    const metric = await getPopulatedMetric();

    reactImmediately(() => {
      expect(metric.singleFollowupDemographics).toMatchSnapshot();
    });

    expect.hasAssertions();
  });

  test.each([["raceOrEthnicity"], ["gender"], ["ageBucket"]] as [
    Exclude<DemographicView, "nofilter">
  ][])("%s data", async (demographicView) => {
    const metric = await getPopulatedMetric();

    runInAction(() => {
      metric.demographicView = demographicView;
    });

    reactImmediately(() => {
      expect(metric.singleFollowupDemographics).toMatchSnapshot();
    });

    expect.hasAssertions();
  });
});
