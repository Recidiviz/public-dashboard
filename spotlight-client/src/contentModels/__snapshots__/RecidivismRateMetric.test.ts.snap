// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`single followup period ageBucket data 1`] = `
Array [
  Object {
    "label": "<25",
    "records": Array [
      Object {
        "color": "#25636F",
        "denominator": 116,
        "label": "2009",
        "pct": 0.39313878244128553,
        "value": 45,
      },
      Object {
        "color": "#25636F",
        "denominator": 117,
        "label": "2010",
        "pct": 0.35398806455471876,
        "value": 41,
      },
      Object {
        "color": "#25636F",
        "denominator": 65,
        "label": "2011",
        "pct": 0.5001636886397268,
        "value": 32,
      },
      Object {
        "color": "#25636F",
        "denominator": 74,
        "label": "2012",
        "pct": 0.3665902040105169,
        "value": 27,
      },
      Object {
        "color": "#25636F",
        "denominator": 77,
        "label": "2013",
        "pct": 0.27464620972057485,
        "value": 21,
      },
      Object {
        "color": "#25636F",
        "denominator": 93,
        "label": "2014",
        "pct": 0.5047399015877124,
        "value": 46,
      },
      Object {
        "color": "#25636F",
        "denominator": 73,
        "label": "2015",
        "pct": 0.4848114314171283,
        "value": 35,
      },
      Object {
        "color": "#25636F",
        "denominator": 118,
        "label": "2016",
        "pct": 0.44484105840224253,
        "value": 52,
      },
    ],
  },
  Object {
    "label": "25-29",
    "records": Array [
      Object {
        "color": "#25636F",
        "denominator": 96,
        "label": "2009",
        "pct": 0.36382424006480313,
        "value": 34,
      },
      Object {
        "color": "#25636F",
        "denominator": 96,
        "label": "2010",
        "pct": 0.4530695537042013,
        "value": 43,
      },
      Object {
        "color": "#25636F",
        "denominator": 53,
        "label": "2011",
        "pct": 0.21644253915033027,
        "value": 11,
      },
      Object {
        "color": "#25636F",
        "denominator": 61,
        "label": "2012",
        "pct": 0.26050703801768876,
        "value": 15,
      },
      Object {
        "color": "#25636F",
        "denominator": 64,
        "label": "2013",
        "pct": 0.34958843720246047,
        "value": 22,
      },
      Object {
        "color": "#25636F",
        "denominator": 76,
        "label": "2014",
        "pct": 0.29669433849450566,
        "value": 22,
      },
      Object {
        "color": "#25636F",
        "denominator": 60,
        "label": "2015",
        "pct": 0.27174598364307445,
        "value": 16,
      },
      Object {
        "color": "#25636F",
        "denominator": 97,
        "label": "2016",
        "pct": 0.23654226546764148,
        "value": 22,
      },
    ],
  },
  Object {
    "label": "30-34",
    "records": Array [
      Object {
        "color": "#25636F",
        "denominator": 117,
        "label": "2009",
        "pct": 0.34592942626462897,
        "value": 40,
      },
      Object {
        "color": "#25636F",
        "denominator": 118,
        "label": "2010",
        "pct": 0.3533961654965064,
        "value": 41,
      },
      Object {
        "color": "#25636F",
        "denominator": 65,
        "label": "2011",
        "pct": 0.3381762354345907,
        "value": 21,
      },
      Object {
        "color": "#25636F",
        "denominator": 75,
        "label": "2012",
        "pct": 0.2997419121789921,
        "value": 22,
      },
      Object {
        "color": "#25636F",
        "denominator": 78,
        "label": "2013",
        "pct": 0.5020914839800488,
        "value": 39,
      },
      Object {
        "color": "#25636F",
        "denominator": 94,
        "label": "2014",
        "pct": 0.43980239410058247,
        "value": 41,
      },
      Object {
        "color": "#25636F",
        "denominator": 74,
        "label": "2015",
        "pct": 0.29413055653070513,
        "value": 21,
      },
      Object {
        "color": "#25636F",
        "denominator": 119,
        "label": "2016",
        "pct": 0.26091572422805687,
        "value": 31,
      },
    ],
  },
  Object {
    "label": "35-39",
    "records": Array [
      Object {
        "color": "#25636F",
        "denominator": 86,
        "label": "2009",
        "pct": 0.2944076241843945,
        "value": 25,
      },
      Object {
        "color": "#25636F",
        "denominator": 87,
        "label": "2010",
        "pct": 0.32430245366805804,
        "value": 28,
      },
      Object {
        "color": "#25636F",
        "denominator": 48,
        "label": "2011",
        "pct": 0.24533156492763689,
        "value": 11,
      },
      Object {
        "color": "#25636F",
        "denominator": 55,
        "label": "2012",
        "pct": 0.46655674830491856,
        "value": 25,
      },
      Object {
        "color": "#25636F",
        "denominator": 58,
        "label": "2013",
        "pct": 0.3595017913313149,
        "value": 20,
      },
      Object {
        "color": "#25636F",
        "denominator": 69,
        "label": "2014",
        "pct": 0.3626761880218852,
        "value": 25,
      },
      Object {
        "color": "#25636F",
        "denominator": 54,
        "label": "2015",
        "pct": 0.3062906781394077,
        "value": 16,
      },
      Object {
        "color": "#25636F",
        "denominator": 88,
        "label": "2016",
        "pct": 0.2708830191249996,
        "value": 23,
      },
    ],
  },
  Object {
    "label": "40<",
    "records": Array [
      Object {
        "color": "#25636F",
        "denominator": 26,
        "label": "2009",
        "pct": 0.37039467193523745,
        "value": 9,
      },
      Object {
        "color": "#25636F",
        "denominator": 27,
        "label": "2010",
        "pct": 0.47831190109530425,
        "value": 12,
      },
      Object {
        "color": "#25636F",
        "denominator": 15,
        "label": "2011",
        "pct": 0.28473022897722605,
        "value": 4,
      },
      Object {
        "color": "#25636F",
        "denominator": 17,
        "label": "2012",
        "pct": 0.33026048060707197,
        "value": 5,
      },
      Object {
        "color": "#25636F",
        "denominator": 18,
        "label": "2013",
        "pct": 0.41270153398843556,
        "value": 7,
      },
      Object {
        "color": "#25636F",
        "denominator": 21,
        "label": "2014",
        "pct": 0.23860421437773077,
        "value": 5,
      },
      Object {
        "color": "#25636F",
        "denominator": 17,
        "label": "2015",
        "pct": 0.3829962083027624,
        "value": 6,
      },
      Object {
        "color": "#25636F",
        "denominator": 27,
        "label": "2016",
        "pct": 0.4379753894586494,
        "value": 11,
      },
    ],
  },
]
`;

exports[`single followup period gender data 1`] = `
Array [
  Object {
    "label": "Male",
    "records": Array [
      Object {
        "color": "#25636F",
        "denominator": 281,
        "label": "2009",
        "pct": 0.24869136439558084,
        "value": 69,
      },
      Object {
        "color": "#25636F",
        "denominator": 283,
        "label": "2010",
        "pct": 0.46398171568994606,
        "value": 131,
      },
      Object {
        "color": "#25636F",
        "denominator": 157,
        "label": "2011",
        "pct": 0.31355400245648035,
        "value": 49,
      },
      Object {
        "color": "#25636F",
        "denominator": 180,
        "label": "2012",
        "pct": 0.3618074406179642,
        "value": 65,
      },
      Object {
        "color": "#25636F",
        "denominator": 187,
        "label": "2013",
        "pct": 0.4266951987202952,
        "value": 79,
      },
      Object {
        "color": "#25636F",
        "denominator": 225,
        "label": "2014",
        "pct": 0.41129634928006986,
        "value": 92,
      },
      Object {
        "color": "#25636F",
        "denominator": 176,
        "label": "2015",
        "pct": 0.3348862947670234,
        "value": 58,
      },
      Object {
        "color": "#25636F",
        "denominator": 286,
        "label": "2016",
        "pct": 0.33309356613108093,
        "value": 95,
      },
    ],
  },
  Object {
    "label": "Female",
    "records": Array [
      Object {
        "color": "#25636F",
        "denominator": 160,
        "label": "2009",
        "pct": 0.3181170286412642,
        "value": 50,
      },
      Object {
        "color": "#25636F",
        "denominator": 161,
        "label": "2010",
        "pct": 0.5837759054688773,
        "value": 93,
      },
      Object {
        "color": "#25636F",
        "denominator": 89,
        "label": "2011",
        "pct": 0.4607570761616222,
        "value": 41,
      },
      Object {
        "color": "#25636F",
        "denominator": 103,
        "label": "2012",
        "pct": 0.3384422015280911,
        "value": 34,
      },
      Object {
        "color": "#25636F",
        "denominator": 107,
        "label": "2013",
        "pct": 0.47601493062591504,
        "value": 50,
      },
      Object {
        "color": "#25636F",
        "denominator": 128,
        "label": "2014",
        "pct": 0.4469079604452117,
        "value": 57,
      },
      Object {
        "color": "#25636F",
        "denominator": 101,
        "label": "2015",
        "pct": 0.4842804591594073,
        "value": 48,
      },
      Object {
        "color": "#25636F",
        "denominator": 163,
        "label": "2016",
        "pct": 0.3463811600337129,
        "value": 56,
      },
    ],
  },
]
`;

exports[`single followup period raceOrEthnicity data 1`] = `
Array [
  Object {
    "label": "Native American",
    "records": Array [
      Object {
        "color": "#25636F",
        "denominator": 147,
        "label": "2009",
        "pct": 0.4310374831525233,
        "value": 63,
      },
      Object {
        "color": "#25636F",
        "denominator": 148,
        "label": "2010",
        "pct": 0.41452789459406847,
        "value": 61,
      },
      Object {
        "color": "#25636F",
        "denominator": 82,
        "label": "2011",
        "pct": 0.33041111630606557,
        "value": 27,
      },
      Object {
        "color": "#25636F",
        "denominator": 94,
        "label": "2012",
        "pct": 0.38790747657753666,
        "value": 36,
      },
      Object {
        "color": "#25636F",
        "denominator": 98,
        "label": "2013",
        "pct": 0.32725349625274136,
        "value": 32,
      },
      Object {
        "color": "#25636F",
        "denominator": 117,
        "label": "2014",
        "pct": 0.3702755342518667,
        "value": 43,
      },
      Object {
        "color": "#25636F",
        "denominator": 92,
        "label": "2015",
        "pct": 0.37373532131252074,
        "value": 34,
      },
      Object {
        "color": "#25636F",
        "denominator": 149,
        "label": "2016",
        "pct": 0.35903743790238474,
        "value": 53,
      },
    ],
  },
  Object {
    "label": "Black",
    "records": Array [
      Object {
        "color": "#25636F",
        "denominator": 38,
        "label": "2009",
        "pct": 0.24093034729343787,
        "value": 9,
      },
      Object {
        "color": "#25636F",
        "denominator": 38,
        "label": "2010",
        "pct": 0.2714481677671729,
        "value": 10,
      },
      Object {
        "color": "#25636F",
        "denominator": 21,
        "label": "2011",
        "pct": 0.22098688455329957,
        "value": 4,
      },
      Object {
        "color": "#25636F",
        "denominator": 25,
        "label": "2012",
        "pct": 0.4064911593897479,
        "value": 10,
      },
      Object {
        "color": "#25636F",
        "denominator": 25,
        "label": "2013",
        "pct": 0.49732304633219726,
        "value": 12,
      },
      Object {
        "color": "#25636F",
        "denominator": 31,
        "label": "2014",
        "pct": 0.3839336630680912,
        "value": 11,
      },
      Object {
        "color": "#25636F",
        "denominator": 24,
        "label": "2015",
        "pct": 0.3299173865620453,
        "value": 7,
      },
      Object {
        "color": "#25636F",
        "denominator": 39,
        "label": "2016",
        "pct": 0.23546230193703543,
        "value": 9,
      },
    ],
  },
  Object {
    "label": "Hispanic",
    "records": Array [
      Object {
        "color": "#25636F",
        "denominator": 100,
        "label": "2009",
        "pct": 0.46066132952357136,
        "value": 46,
      },
      Object {
        "color": "#25636F",
        "denominator": 101,
        "label": "2010",
        "pct": 0.4213290399574359,
        "value": 42,
      },
      Object {
        "color": "#25636F",
        "denominator": 56,
        "label": "2011",
        "pct": 0.44386132162712605,
        "value": 24,
      },
      Object {
        "color": "#25636F",
        "denominator": 64,
        "label": "2012",
        "pct": 0.2168935769006367,
        "value": 13,
      },
      Object {
        "color": "#25636F",
        "denominator": 67,
        "label": "2013",
        "pct": 0.3436280912901396,
        "value": 23,
      },
      Object {
        "color": "#25636F",
        "denominator": 80,
        "label": "2014",
        "pct": 0.46971296647420574,
        "value": 37,
      },
      Object {
        "color": "#25636F",
        "denominator": 63,
        "label": "2015",
        "pct": 0.3895765803300314,
        "value": 24,
      },
      Object {
        "color": "#25636F",
        "denominator": 102,
        "label": "2016",
        "pct": 0.3737065125769247,
        "value": 38,
      },
    ],
  },
  Object {
    "label": "White",
    "records": Array [
      Object {
        "color": "#25636F",
        "denominator": 35,
        "label": "2009",
        "pct": 0.23480046292399714,
        "value": 8,
      },
      Object {
        "color": "#25636F",
        "denominator": 35,
        "label": "2010",
        "pct": 0.25360460340051105,
        "value": 8,
      },
      Object {
        "color": "#25636F",
        "denominator": 20,
        "label": "2011",
        "pct": 0.19130527518702728,
        "value": 3,
      },
      Object {
        "color": "#25636F",
        "denominator": 23,
        "label": "2012",
        "pct": 0.5098740144908329,
        "value": 11,
      },
      Object {
        "color": "#25636F",
        "denominator": 23,
        "label": "2013",
        "pct": 0.6092384625296211,
        "value": 14,
      },
      Object {
        "color": "#25636F",
        "denominator": 28,
        "label": "2014",
        "pct": 0.3294634564258997,
        "value": 9,
      },
      Object {
        "color": "#25636F",
        "denominator": 22,
        "label": "2015",
        "pct": 0.23629892983690073,
        "value": 5,
      },
      Object {
        "color": "#25636F",
        "denominator": 36,
        "label": "2016",
        "pct": 0.37767495725981837,
        "value": 13,
      },
    ],
  },
  Object {
    "label": "Other",
    "records": Array [
      Object {
        "color": "#25636F",
        "denominator": 121,
        "label": "2009",
        "pct": 0.45046609809901517,
        "value": 54,
      },
      Object {
        "color": "#25636F",
        "denominator": 122,
        "label": "2010",
        "pct": 0.3078691384827722,
        "value": 37,
      },
      Object {
        "color": "#25636F",
        "denominator": 67,
        "label": "2011",
        "pct": 0.25894059644135314,
        "value": 17,
      },
      Object {
        "color": "#25636F",
        "denominator": 78,
        "label": "2012",
        "pct": 0.23323275706861413,
        "value": 18,
      },
      Object {
        "color": "#25636F",
        "denominator": 81,
        "label": "2013",
        "pct": 0.4004864028444184,
        "value": 32,
      },
      Object {
        "color": "#25636F",
        "denominator": 97,
        "label": "2014",
        "pct": 0.4043962112906952,
        "value": 39,
      },
      Object {
        "color": "#25636F",
        "denominator": 76,
        "label": "2015",
        "pct": 0.38997840956410185,
        "value": 29,
      },
      Object {
        "color": "#25636F",
        "denominator": 123,
        "label": "2016",
        "pct": 0.3140820281613471,
        "value": 38,
      },
    ],
  },
]
`;

exports[`single followup period total data 1`] = `
Array [
  Object {
    "label": "Total",
    "records": Array [
      Object {
        "color": "#25636F",
        "denominator": 441,
        "label": "2009",
        "pct": 0.27481568501675263,
        "value": 121,
      },
      Object {
        "color": "#25636F",
        "denominator": 444,
        "label": "2010",
        "pct": 0.5705945384276757,
        "value": 253,
      },
      Object {
        "color": "#25636F",
        "denominator": 246,
        "label": "2011",
        "pct": 0.3865270115208075,
        "value": 95,
      },
      Object {
        "color": "#25636F",
        "denominator": 283,
        "label": "2012",
        "pct": 0.38288543842088013,
        "value": 108,
      },
      Object {
        "color": "#25636F",
        "denominator": 294,
        "label": "2013",
        "pct": 0.3966145512517424,
        "value": 116,
      },
      Object {
        "color": "#25636F",
        "denominator": 353,
        "label": "2014",
        "pct": 0.3704884876536363,
        "value": 130,
      },
      Object {
        "color": "#25636F",
        "denominator": 277,
        "label": "2015",
        "pct": 0.30393467040023203,
        "value": 84,
      },
      Object {
        "color": "#25636F",
        "denominator": 449,
        "label": "2016",
        "pct": 0.20221756415560455,
        "value": 90,
      },
    ],
  },
]
`;
